### PR TITLE
Implement tangent and secant planes for transcendental functions

### LINKED ISSUE #1168: Support tangent planes for transcendental functions
Currently our candidate models for problems involving transcendental functions are very imprecise.

We should implement Figure 3 of https://es-static.fbk.eu/people/griggio/papers/cade17.pdf by Cimatti et al.

This functionality has been requested by GE.

### PR DESCRIPTION
**Pull request by ajreynol**
This implements Figure 3 of https://es-static.fbk.eu/people/griggio/papers/cade17.pdf, and fixes #1168.

### CODE REVIEW

File: src/options/arith_options [hunk 1]
```diff
@@ -176,6 +176,9 @@ option nlExtFactor --nl-ext-factor bool :default true
  
 option nlExtTangentPlanes --nl-ext-tplanes bool :default false
  use non-terminating tangent plane strategy for non-linear
+
```
Comment (HanielB) [inline src/options/arith_options]: trailing space

File: src/theory/arith/nonlinear_extension.h [hunk 2]
```diff
@@ -705,6 +705,64 @@ class NonlinearExtension {
   */
   std::vector<Node> checkTranscendentalMonotonic();
 
+  /** check transcendental tangent planes
+  *
+  * Returns a set of valid theory lemmas, based on
+  * computing a "incremental linearization" of
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: a --> an

File: src/theory/arith/nonlinear_extension.h [hunk 3]
```diff
@@ -705,6 +705,64 @@ class NonlinearExtension {
...
+  * It is based on Figure 3 of "Satisfiability
+  * Modulo Transcendental Functions via Incremental
+  * Linearization" by Cimatti et al., CADE 2017.
+  * This schema is not terminating in general.
+  * It is not enabled by default, and can
+  * be enabled by --nl-ext-tf-tplanes.
+  *
+  * Example:
+  *
+  * Assume we have a term sin(y) where y^M = 1, note that
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: Elsewhere you use M() for interpretation. Change to that.

File: src/theory/arith/nonlinear_extension.cpp [hunk 4]
```diff
@@ -2910,6 +2920,439 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
   return lemmas;
 }
 
+std::vector<Node> NonlinearExtension::checkTranscendentalTangentPlanes()
+{
+  std::vector<Node> lemmas;
+  Trace("nl-ext") << "Get tangent plane lemmas for transcendental functions..."
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: for-each

File: src/theory/arith/nonlinear_extension.cpp [hunk 5]
```diff
@@ -2910,6 +2920,439 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+
+  // this implements Figure 3 of "Satisfiaility Modulo Transcendental Functions
+  // via Incremental Linearization" by Cimatti et al
+  for (std::map<Kind, std::map<Node, Node> >::iterator it =
+           d_tf_rep_map.begin();
+       it != d_tf_rep_map.end();
+       ++it)
+  {
+    Kind k = it->first;
+    Node tft = NodeManager::currentNM()->mkNode(k, d_zero);
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: cache ``NodeManager::currentNM()``

File: src/theory/arith/nonlinear_extension.cpp [hunk 6]
```diff
@@ -2910,6 +2920,439 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
 
+std::vector<Node> NonlinearExtension::checkTranscendentalTangentPlanes()
+{
+  std::vector<Node> lemmas;
+  Trace("nl-ext") << "Get tangent plane lemmas for transcendental functions..."
+                  << std::endl;
+
+  // this implements Figure 3 of "Satisfiaility Modulo Transcendental Functions
+  // via Incremental Linearization" by Cimatti et al
+  for (std::map<Kind, std::map<Node, Node> >::iterator it =
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: You are using the namespace std and CVC4::kind in this file. Either remove them or remove the prefixes.

File: src/theory/arith/nonlinear_extension.cpp [hunk 7]
```diff
@@ -2910,6 +2920,439 @@ std::vector<Node> NonlinearExtension::checkTranscendentalMonotonic() {
...
+    Trace("nl-ext-tf-tplanes") << " Lower (pos): " << poly_approx_bounds[0][1]
+                               << std::endl;
+    Trace("nl-ext-tf-tplanes") << " Upper (pos): " << poly_approx_bounds[1][1]
+                               << std::endl;
+    Trace("nl-ext-tf-tplanes") << " Lower (neg): " << poly_approx_bounds[0][-1]
+                               << std::endl;
+    Trace("nl-ext-tf-tplanes") << " Upper (neg): " << poly_approx_bounds[1][-1]
+                               << std::endl;
+
+    for (std::map<Node, Node>::iterator it2 = it->second.begin();
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: for-each

**Comment by ajreynol** [comment]
I've addressed the review, also removed all kind:: from the .cpp file.

**Pull request merged**